cnd/src/network/behaviour.rs

```diff
@@ -0,0 +1,160 @@
+use crate::network::protocol::{InboundProtocolConfig, Message, OutboundProtocolConfig};
+use libp2p::{
+    core::{ConnectedPoint, Multiaddr, PeerId},
+    swarm::{
+        NetworkBehaviour, NetworkBehaviourAction, OneShotHandler, PollParameters, ProtocolsHandler,
+    },
+};
+use std::{
+    collections::VecDeque,
+    task::{Context, Poll},
+};
+
+/// Network behaviour that handles the secret hash protocol.
+#[derive(Debug)]
+pub struct Behaviour {
+    /// Events that need to be yielded to the outside when polling.
+    events: VecDeque<NetworkBehaviourAction<OutboundProtocolConfig, OutEvent>>,
+}
+
+/// Event generated by the NetworkBehaviour and that the swarm will report back.
+#[derive(Clone, Copy, Debug)]
+pub enum OutEvent {
+    Rx(Message), // OutEvent containing secret hash message.
+    Sent,        // Empty/nil OutEvent i.e., `()`.
```

Comment:
Who is generating the `Sent` event, do we really need this?

Reply:
I'll be damned if I can work out when events get generated but from what I can tell this event gets generated when we are the dialer and we do the OutboundUpgrade ???  At least the compiler requires us to have this and also the From<()> impl.  Also the Floodsub protocol as this like this also.  Please do school me if you work this out.